src/core/ddsi/src/ddsi_serdata_cdr.c

```diff
@@ -0,0 +1,361 @@
+  d->pos = 0;
+  d->hdr.identifier = ddsi_sertype_get_native_enc_identifier (DDS_CDR_ENC_VERSION_2, tp->encoding_format);
+  d->hdr.options = 0;
+}
+
+static struct ddsi_serdata_cdr *serdata_cdr_allocnew (uint32_t init_size)
+{
+  struct ddsi_serdata_cdr *d = ddsrt_malloc (offsetof (struct ddsi_serdata_cdr, data) + init_size);
+  d->size = init_size;
+  return d;
+}
+
+static struct ddsi_serdata_cdr *serdata_cdr_new_size (const struct ddsi_sertype_cdr *tp, enum ddsi_serdata_kind kind, uint32_t size)
+{
+  struct ddsi_serdata_cdr *d;
+  if ((d = serdata_cdr_allocnew (size)) == NULL)
+    return NULL;
+  serdata_cdr_init (d, tp, kind);
+  return d;
+}
+
+static struct ddsi_serdata_cdr *serdata_cdr_new (const struct ddsi_sertype_cdr *tp, enum ddsi_serdata_kind kind)
+{
+  return serdata_cdr_new_size (tp, kind, DEFAULT_NEW_SIZE);
+}
+
+static inline bool is_valid_xcdr_id (unsigned short cdr_identifier)
+{
+  return (cdr_identifier == CDR2_LE || cdr_identifier == CDR2_BE
+    || cdr_identifier == D_CDR2_LE || cdr_identifier == D_CDR2_BE
+    || cdr_identifier == PL_CDR2_LE || cdr_identifier == PL_CDR2_BE);
+}
+
+/* Construct a serdata from a fragchain received over the network */
+static struct ddsi_serdata_cdr *serdata_cdr_from_ser_common (const struct ddsi_sertype *tpcmn, enum ddsi_serdata_kind kind, const struct nn_rdata *fragchain, size_t size)
+{
+  assert (kind == SDK_DATA);
+  const struct ddsi_sertype_cdr *tp = (const struct ddsi_sertype_cdr *)tpcmn;
+
+  /* FIXME: check whether this really is the correct maximum: offsets are relative
+     to the CDR header, but there are also some places that use a serdata as-if it
+     were a stream, and those use offsets (m_index) relative to the start of the
+     serdata */
+  if (size > UINT32_MAX - offsetof (struct ddsi_serdata_cdr, hdr))
+    return NULL;
+  struct ddsi_serdata_cdr *d = serdata_cdr_new_size (tp, kind, (uint32_t) size);
+  if (d == NULL)
+    return NULL;
+
+  uint32_t off = 4; /* must skip the CDR header */
+
+  assert (fragchain->min == 0);
+  assert (fragchain->maxp1 >= off); /* CDR header must be in first fragment */
+
+  memcpy (&d->hdr, NN_RMSG_PAYLOADOFF (fragchain->rmsg, NN_RDATA_PAYLOAD_OFF (fragchain)), sizeof (d->hdr));
+  if (!is_valid_xcdr_id (d->hdr.identifier))
+    goto err;
+
+  while (fragchain)
+  {
+    assert (fragchain->min <= off);
+    assert (fragchain->maxp1 <= size);
+    if (fragchain->maxp1 > off)
+    {
+      /* only copy if this fragment adds data */
+      const unsigned char *payload = NN_RMSG_PAYLOADOFF (fragchain->rmsg, NN_RDATA_PAYLOAD_OFF (fragchain));
+      serdata_cdr_append_blob (&d, fragchain->maxp1 - off, payload + off - fragchain->min);
+      off = fragchain->maxp1;
+    }
+    fragchain = fragchain->nextfrag;
+  }
+
+  const bool needs_bswap = !CDR_ENC_IS_NATIVE (d->hdr.identifier);
+  d->hdr.identifier = CDR_ENC_TO_NATIVE (d->hdr.identifier);
+  const uint32_t pad = ddsrt_fromBE2u (d->hdr.options) & 2;
```

Comment:
This `& 2` is weird. It was there before in `serdata_default`, but either that was in error and it should be `& 3` (and we are missing a test case 🤓) or it needs a comment why it is correct!

Reply:
Good catch! Fixed in [616f7c4](https://github.com/eclipse-cyclonedds/cyclonedds/pull/1468/commits/616f7c4b2a12f4a03301239d49397d68d6868dff)